### DISCUSSION TITLE
Encoders and MotorEncoders readings

### DISCUSSION THREAD

**Discussion by hu-yue** [robotology/community#53] [Q&A] [answered]
Hi,
I used iEncoders and iMotorEncoders to read the encoders of the robot, as I wanted to test the effect of the springs in the joints.
However, I'm not sure about how to interpret the readings from iMotorEncoders. I thought it is enough to divide by the gear ration (100), but the numbers I obtained are not the same as iEncoders even when the springs are not mounted.
And as I've understood, iEncoders and iMotorEncoders should read the same values (almost?) when no spring is mounted.

What is the correct way to interpret the readings of iMotorEncoders?
@DanielePucci was also interested in this.

Thanks!

**Comment by francesco-romano** [comment]
Have you tried to see what  [getGearboxRatio](http://wiki.icub.org/yarpdoc/classyarp_1_1dev_1_1IMotor.html#a6abbd0ad79a61b17872f4c0f513654e7) returns?

**Comment by hu-yue** [comment]
Nope, can try that.

**Comment by randaz81** [comment, marked as answer]
Yes, indeed you have to divide the position returned by iMotorEncoders by the gear ratio obtained by getGearBoxRatio, considering also the sign. Additionally, you have to remove the offset which affects the motor encoder (indeed they are relative encoder, not absolute encoders). check `opticalEncoders-consistency` in `icub-tests` for an example

**Comment by hu-yue** [comment]
I get false from getGearboxRatio and so nonsense gear ratio, is the integer number m of the arguments ranging from 0 to 5 for example in the case of left_leg? 

```
getGearboxRatio (int m, double *val)
```

**Comment by randaz81** [comment]
yes, it is. For the legs legs it should return *val= +-100.
If the interface method return false, one possibility is that your console laptop/pc104/firmware(?) is too old and needs an update... (after IROS)

**Comment by hu-yue** [comment]
Ok, but it return false and val = 0. And also the function `getNumberOfMotors (int *num)` returns false and num is 0. 
But this part did not give any error:

```
    Property options;
    options.put("device", "remote_controlboard");
    options.put("local", localPorts.c_str());
    options.put("remote", remotePorts.c_str());
    PolyDriver robotDevice(options);
    IMotor *motors;
    if (!robotDevice.view(motors)) {
        printf("Problems acquiring interfaces\n");
        return 0;
    }
```

And the iEncoders and iMotorEncoders seem to work fine.

**Comment by randaz81** [comment]
Looking at git history, icub-main has be updated at least at 04 July to properly manage getGearboxRatio. Is it so?

**Comment by hu-yue** [comment]
icub-main is updated up to 04/09/2015, as well as yarp. But the firmware is from June.

**Comment by pattacini** [comment]
@hu-yue any progress on this?

**Comment by hu-yue** [comment]
@pattacini not yet, sorry, I just got back from Hamburg but have to leave for Delft already so I did not have time to check the firmware issue with @randaz81, maybe we can do it next week?

**Comment by pattacini** [comment]
there's no hurry, was just a ping :smirk:

**Comment by hu-yue** [comment]
After the last firmware and software update now it's possible to read the gear ratio using `getGearboxRatio (int m, double *val)`.
The motor encoders values were converted into encoders values as
`(m_encoders - enc_offsets)/gear_ratio`
The difference between m_encoders and encoders reading are of order 10^-2, is this reasonable?

**Comment by randaz81** [comment]
yes 10^-2 deg is ok. 
btw, the module `opticalEncoders-consistency` has been recently renamed to `motorEncoders-consistency`. This test is more complete in the sense that checks also for the consistency between the joint velocity and the motor velocity.

**Comment by hu-yue** [comment]
Ok thanks!

**Comment by traversaro** [comment]
fyi @fjandrad
